File: tools/core070_aghq_batch.jl
# Inner executor for the "aghq control" M2 batch (16 paired-control cases,
# all decided here; see docs/dev-log/core070/aghq-batch-contract.json for the
# full case list and the 22 NEEDS_NEW_JULIA_SURFACE deferrals).
#
# Pure-Julia consumer, mirroring tools/core070_namespace_2_batch.jl: reads a
# JSON oracle file the paired R runner (tools/core070_aghq_batch.R) writes
# BEFORE invoking this script (that R process already has the frozen
# gllvmTMB library loaded and evaluates every frozen r_call/r_assertion pair
# itself), then evaluates the identical GLLVM._aghq_request(...) calls
# natively via direct `using GLLVM` module calls only. No RCall, no parity-
# runner include, no R of any kind runs in this process. No model fit
# anywhere: GLLVM._aghq_request is a pure scalar-argument normalizer.
#
# argv:
#   ARGS[1]  destination path for the JSON results file (parent dir need not
#            exist yet; this script mkpath()s it). Must not already exist.
#
# Env vars required (set by the outer R runner before invoking this script;
# never defaulted silently here so a misconfigured environment fails loudly
# rather than quietly running against stale or missing oracle data):
#   CORE070_AGHQ_R_ORACLE = <path to the R-written oracle JSON>
#
# Invocation:
#   julia --project=. tools/core070_aghq_batch.jl <out.json>

using GLLVM

# ---------------------------------------------------------------------------
# Minimal JSON reader/writer (no external dependency; mirrors the existing
# repo convention in tools/core070_namespace_2_batch.jl).
# ---------------------------------------------------------------------------
function json_read(path::AbstractString)
    txt = read(path, String)
    pos = Ref(1)
    skip_ws!() = begin
        while pos[] <= lastindex(txt) && isspace(txt[pos[]])
            pos[] = nextind(txt, pos[])
        end
    end
    local parse_value
    function parse_string()
        pos[] += 1 # opening quote
        buf = IOBuffer()
        while true
            c = txt[pos[]]
            if c == '"'
                pos[] = nextind(txt, pos[])
                break
            elseif c == '\\'
                pos[] = nextind(txt, pos[])
                e = txt[pos[]]
                if e == 'n'; write(buf, '\n')
                elseif e == 't'; write(buf, '\t')
                elseif e == 'u'
                    hex = txt[nextind(txt, pos[]):nextind(txt, pos[], 4)]
                    write(buf, Char(parse(UInt32, hex; base = 16)))
                    pos[] = nextind(txt, pos[], 4)
                else write(buf, e)
                end
                pos[] = nextind(txt, pos[])
            else
                write(buf, c)
                pos[] = nextind(txt, pos[])
            end
        end
        String(take!(buf))
    end
    function parse_number()
        start = pos[]
        while pos[] <= lastindex(txt) && (isdigit(txt[pos[]]) || txt[pos[]] in ('-', '+', '.', 'e', 'E'))
            pos[] = nextind(txt, pos[])
        end
        s = txt[start:prevind(txt, pos[])]
        return occursin('.', s) || occursin('e', s) || occursin('E', s) ? parse(Float64, s) : parse(Int, s)
    end
    function parse_array()
        pos[] += 1
        out = Any[]
        skip_ws!()
        if txt[pos[]] == ']'
            pos[] = nextind(txt, pos[])
            return out
        end
        while true
            skip_ws!()
            push!(out, parse_value())
            skip_ws!()
            if txt[pos[]] == ','
                pos[] = nextind(txt, pos[])
            elseif txt[pos[]] == ']'
                pos[] = nextind(txt, pos[])
                break
            end
        end
        out
    end
    function parse_object()
        pos[] += 1
        out = Dict{String, Any}()
        skip_ws!()
        if txt[pos[]] == '}'
            pos[] = nextind(txt, pos[])
            return out
        end
        while true
            skip_ws!()
            k = parse_string()
            skip_ws!()
            @assert txt[pos[]] == ':'
            pos[] = nextind(txt, pos[])
            skip_ws!()
            out[k] = parse_value()
            skip_ws!()
            if txt[pos[]] == ','
                pos[] = nextind(txt, pos[])
            elseif txt[pos[]] == '}'
                pos[] = nextind(txt, pos[])
                break
            end
        end
        out
    end
    parse_value = function ()
        skip_ws!()
        c = txt[pos[]]
        if c == '"'; return parse_string()
        elseif c == '['; return parse_array()
        elseif c == '{'; return parse_object()
        elseif c == 't'; pos[] += 4; return true
        elseif c == 'f'; pos[] += 5; return false
        elseif c == 'n'; pos[] += 4; return nothing
        else; return parse_number()
        end
    end
    skip_ws!()
    parse_value()
end

json_escape(s::AbstractString) = replace(s, "\\" => "\\\\", "\"" => "\\\"", "\n" => "\\n")
to_json(x::Bool) = x ? "true" : "false"
to_json(x::Nothing) = "null"
to_json(x::AbstractString) = "\"" * json_escape(x) * "\""
to_json(x::Integer) = string(x)
to_json(x::AbstractFloat) = isfinite(x) ? repr(x) : "null"
to_json(x::AbstractVector) = "[" * join(to_json.(x), ",") * "]"
to_json(x::AbstractDict) = "{" * join(("\"$(json_escape(string(k)))\":" * to_json(v) for (k, v) in x), ",") * "}"

length(ARGS) == 1 || error("usage: julia tools/core070_aghq_batch.jl <out.json>")
out_path = ARGS[1]
isfile(out_path) && error("destination already exists: $out_path")
mkpath(dirname(out_path))

@assert realpath(Base.pkgdir(GLLVM)) == realpath(pwd()) "must run from the GLLVM.jl package root"

oracle_path = get(ENV, "CORE070_AGHQ_R_ORACLE", "")
isempty(oracle_path) &&
    error("CORE070_AGHQ_R_ORACLE is required; refusing to silently default it")
isfile(oracle_path) ||
    error("CORE070_AGHQ_R_ORACLE points to a nonexistent file: $oracle_path")

oracle = json_read(oracle_path)
r_cases = oracle["cases"]

# ---------------------------------------------------------------------------
# Native normalization label helper: mirrors expected strings from the
# frozen contract ("off", "auto", "1"/"2"/"9", "ArgumentError").
# ---------------------------------------------------------------------------
function _label(v)
    v === :off && return "off"
    v === :auto && return "auto"
    v isa Integer && return string(v)
    return "unexpected:$(repr(v))"
end

function _run(f)
    try
        return (label = _label(f()), errored = false)
    catch e
        e isa ArgumentError && return (label = "ArgumentError", errored = true)
        return (label = "unexpected_exception:$(sprint(showerror, e))", errored = true)
    end
end

# ---------------------------------------------------------------------------
# The 16 paired-control cases, literally matching each case's `julia_call`
# in docs/dev-log/core070/aghq-batch-contract.json.
# ---------------------------------------------------------------------------
case_specs = [
    ("CORE070-AGHQ-CTRL-FALSE-PAIRED-CONTROL",    () -> GLLVM._aghq_request(false),   "off"),
    ("CORE070-AGHQ-CTRL-NULL-PAIRED-CONTROL",     () -> GLLVM._aghq_request(nothing), "off"),
    ("CORE070-AGHQ-CTRL-TRUE-PAIRED-CONTROL",     () -> GLLVM._aghq_request(true),    "auto"),
    ("CORE070-AGHQ-CTRL-AUTO-PAIRED-CONTROL",     () -> GLLVM._aghq_request(:auto),   "auto"),
    ("CORE070-AGHQ-CTRL-ONE-PAIRED-CONTROL",      () -> GLLVM._aghq_request(1),       "1"),
    ("CORE070-AGHQ-CTRL-TWO-PAIRED-CONTROL",      () -> GLLVM._aghq_request(2),       "2"),
    ("CORE070-AGHQ-CTRL-NINE-PAIRED-CONTROL",     () -> GLLVM._aghq_request(9),       "9"),
    ("CORE070-AGHQ-INVALID-ZERO-PAIRED-CONTROL",     () -> GLLVM._aghq_request(0),      "ArgumentError"),
    ("CORE070-AGHQ-INVALID-NEGATIVE-PAIRED-CONTROL", () -> GLLVM._aghq_request(-1),     "ArgumentError"),
    ("CORE070-AGHQ-INVALID-FRACTION-PAIRED-CONTROL", () -> GLLVM._aghq_request(1.5),    "ArgumentError"),
    ("CORE070-AGHQ-INVALID-INF-PAIRED-CONTROL",       () -> GLLVM._aghq_request(Inf),    "ArgumentError"),
    ("CORE070-AGHQ-INVALID-NAN-PAIRED-CONTROL",       () -> GLLVM._aghq_request(NaN),    "ArgumentError"),
    ("CORE070-AGHQ-INVALID-NA-PAIRED-CONTROL",        () -> GLLVM._aghq_request(missing),"ArgumentError"),
    ("CORE070-AGHQ-INVALID-VECTOR-PAIRED-CONTROL",    () -> GLLVM._aghq_request([3, 5]), "ArgumentError"),
    ("CORE070-AGHQ-INVALID-EMPTY-PAIRED-CONTROL",     () -> GLLVM._aghq_request(Int[]),  "ArgumentError"),
    ("CORE070-AGHQ-INVALID-STRING-PAIRED-CONTROL",    () -> GLLVM._aghq_request("9"),    "ArgumentError"),
]

cases = Dict{String, Any}()
run_result = Dict{String, Any}()  # case_id -> (label, errored) for negative-control reuse below
for (cid, f, expected) in case_specs
    r = _run(f)
    run_result[cid] = r
    r_pass = haskey(r_cases, cid) && r_cases[cid]["r_assertion_pass"] === true
    julia_pass = r.label == expected
    cases[cid] = Dict(
        "pass" => r_pass && julia_pass,
        "r_assertion_pass" => r_pass,
        "julia_label" => r.label,
        "expected" => expected,
        "julia_pass" => julia_pass,
    )
end

# ---------------------------------------------------------------------------
# Negative controls: deliberately-wrong comparisons that MUST fail.
# ---------------------------------------------------------------------------
false_label = run_result["CORE070-AGHQ-CTRL-FALSE-PAIRED-CONTROL"].label   # "off"
one_label   = run_result["CORE070-AGHQ-CTRL-ONE-PAIRED-CONTROL"].label     # "1"
zero_label  = run_result["CORE070-AGHQ-INVALID-ZERO-PAIRED-CONTROL"].label # "ArgumentError"

neg_flipped = false_label == "auto"        # must be false: "off" != "auto"
neg_valid_as_error = one_label == "ArgumentError"  # must be false: "1" != "ArgumentError"
neg_invalid_as_value = zero_label == "off"         # must be false: "ArgumentError" != "off"

negative_controls = Dict(
    "NEG-EXPECTED-VALUE-FLIPPED" => Dict(
        "behaved" => !neg_flipped, "mismatch_detected" => !neg_flipped),
    "NEG-VALID-CASE-MISCLASSIFIED-AS-ERROR" => Dict(
        "behaved" => !neg_valid_as_error, "mismatch_detected" => !neg_valid_as_error),
    "NEG-INVALID-CASE-MISCLASSIFIED-AS-VALUE" => Dict(
        "behaved" => !neg_invalid_as_value, "mismatch_detected" => !neg_invalid_as_value),
)

all_positive_pass = all(v["pass"] for v in values(cases))
negatives_behaved = all(v["behaved"] for v in values(negative_controls))
overall_ok = all_positive_pass && negatives_behaved

report = Dict(
    "status" => overall_ok ? "PASS" : "FAIL",
    "area" => "aghq-control",
    "scope" => "CORE070_AGHQ_CONTROL_BATCH",
    "case_count" => length(cases),
    "negative_control_count" => length(negative_controls),
    "all_positive_pass" => all_positive_pass,
    "negative_controls_behaved_as_expected" => negatives_behaved,
    "all_checks" => overall_ok,
    "cases" => cases,
    "negative_controls" => negative_controls,
    "julia_version" => string(VERSION),
)
open(out_path, "w") do io
    print(io, to_json(report))
end
println("CORE070_AGHQ_CONTROL_BATCH_RESULT ",
        overall_ok ? "PASS" : "FAIL",
        " positive=", all_positive_pass, " negatives=", negatives_behaved)
overall_ok || exit(1)

File: tools/core070_aghq_batch.R
# Retained evidence runner for the "aghq control" M2 batch (16 paired-control
# cases: 7 executable_case_id-normalization controls (.gllvmTMB_normalize_aghq
# FALSE/NULL/TRUE/"auto"/1/2/9) + 9 invalid-input rejection controls; see
# docs/dev-log/core070/aghq-batch-contract.json for the full case list and
# the 22 NEEDS_NEW_JULIA_SURFACE deferrals carried over from
# docs/dev-log/core070/aghq-required-case-plan.json's source_rows.
#
# Design mirrors tools/core070_namespace_2_batch.R's repair pattern: THIS R
# process (the one with the frozen gllvmTMB library loaded) does 100% of the
# live R-side computation itself (evaluating each frozen r_call/r_assertion
# pair verbatim) and hands the Julia child a plain JSON oracle file; the
# Julia child (tools/core070_aghq_batch.jl) runs zero R code and carries no
# RCall dependency at all.
#
# No model fit anywhere in this batch: .gllvmTMB_normalize_aghq and
# GLLVM._aghq_request are pure scalar-argument normalizers with no data
# dependency, so this whole batch runs in well under a minute.
#
# argv (matches the contract's runner.outer_argv and the --self-test local
# smoke, which never reaches this file):
#   Rscript --vanilla tools/core070_aghq_batch.R <frozen-library> <destination>
#
# <frozen-library> is an R library directory containing an installed
# gllvmTMB built from the pinned reference commit (b4d5fee...) -- the FROZEN,
# INSTALLED library, matching masks_known.R's arg 1, not a source tree.
# <destination> must not already exist; it is created and holds
# r-oracle.json, julia-results.json, julia-stdout.log, julia-stderr.log,
# results.tsv, diagnostics.log, and receipt.json.

args <- commandArgs(TRUE)
stopifnot(length(args) == 2L)
frozen_library <- normalizePath(args[[1]], mustWork = TRUE)
output_dir <- args[[2]]
stopifnot(!dir.exists(output_dir))
dir.create(output_dir, recursive = TRUE)

sha256_file <- function(path) {
  command <- if (nzchar(Sys.which("sha256sum"))) "sha256sum" else "shasum"
  argv <- if (identical(command, "sha256sum")) path else c("-a", "256", path)
  line <- system2(command, argv, stdout = TRUE, stderr = TRUE)
  stopifnot(is.null(attr(line, "status")), length(line) >= 1L)
  sub("[[:space:]].*$", "", line[[1L]])
}

.libPaths(c(frozen_library, .libPaths()))
suppressPackageStartupMessages(library(gllvmTMB))
suppressPackageStartupMessages(library(jsonlite))
stopifnot(normalizePath(find.package("gllvmTMB")) ==
          normalizePath(file.path(frozen_library, "gllvmTMB")))

# .gllvmTMB_normalize_aghq (R/gllvmTMB.R:2046) is unexported, so bare-name
# eval() of the contract's verbatim r_call/r_assertion strings (which call it
# unqualified, matching docs/dev-log/core070/aghq-required-case-plan.json's
# own frozen paired_control_cases text exactly) needs the internal binding
# pulled into the evaluation environment once, up front.
assign(".gllvmTMB_normalize_aghq",
       get(".gllvmTMB_normalize_aghq", envir = asNamespace("gllvmTMB")),
       envir = globalenv())

root <- normalizePath(".")
contract_path <- file.path(root, "docs/dev-log/core070/aghq-batch-contract.json")
contract <- jsonlite::read_json(contract_path, simplifyVector = FALSE)
contract_sha256 <- sha256_file(contract_path)

stopifnot(identical(contract$reference_commit, "b4d5fee64def88bc768dda1f1f77c29b295edd86"),
          identical(contract$status, "FROZEN_AGHQ_CONTROL_BATCH_CONTRACT"),
          length(contract$cases) == contract$expected_case_count,
          contract$expected_case_count == 16L,
          length(contract$negative_controls) >= 3L)

# --- validate pinned R source ------------------------------------------
source_root <- file.path(root, ".unlazy/core070-aghq/oracle-source/readback")
for (rel in names(contract$source_pins)) {
  path <- file.path(source_root, rel)
  stopifnot(file.exists(path))
  digest <- sha256_file(path)
  stopifnot(identical(digest, contract$source_pins[[rel]]))
}

# ---------------------------------------------------------------------------
# 1. Evaluate the frozen r_call / r_assertion pair for each of the 16
#    paired-control cases exactly as written in the contract (verbatim R
#    source text from docs/dev-log/core070/aghq-required-case-plan.json's
#    paired_control_cases; not re-derived).
# ---------------------------------------------------------------------------
r_case_results <- list()
for (case in contract$cases) {
  cid <- case$case_id
  r_value <- tryCatch(
    list(ok = TRUE, value = eval(parse(text = case$r_call)), error = NA_character_),
    error = function(e) list(ok = FALSE, value = NULL, error = conditionMessage(e))
  )
  r_assertion_pass <- tryCatch(
    isTRUE(eval(parse(text = case$r_assertion))),
    error = function(e) FALSE
  )
  r_case_results[[cid]] <- list(
    r_assertion_pass = r_assertion_pass,
    r_call_errored = !r_value$ok,
    r_call_error = if (r_value$ok) NA_character_ else r_value$error
  )
}
stopifnot(all(vapply(r_case_results, function(x) isTRUE(x$r_assertion_pass), logical(1))))

# ---------------------------------------------------------------------------
# 2. Write the oracle JSON (per-case R assertion outcome only -- no model
#    data of any kind) for the Julia child.
# ---------------------------------------------------------------------------
oracle_path <- file.path(output_dir, "r-oracle.json")
jsonlite::write_json(
  list(
    schema = "core070-aghq-r-oracle/v1",
    cases = r_case_results
  ),
  oracle_path, auto_unbox = TRUE, pretty = TRUE, null = "null", na = "null", digits = 17
)

# --- invoke the Julia child ---------------------------------------------
julia_out <- file.path(output_dir, "julia-results.json")
julia_env <- c(CORE070_AGHQ_R_ORACLE = oracle_path)
julia_bin <- Sys.which("julia")
stopifnot(nzchar(julia_bin))
old_env <- Sys.getenv(names(julia_env), unset = NA, names = TRUE)
do.call(Sys.setenv, as.list(julia_env))
t0 <- Sys.time()
julia_status <- tryCatch(
  system2(julia_bin, c("--project=.", "tools/core070_aghq_batch.jl", julia_out),
          stdout = file.path(output_dir, "julia-stdout.log"),
          stderr = file.path(output_dir, "julia-stderr.log")),
  finally = {
    unset <- names(old_env)[is.na(old_env)]
    if (length(unset)) Sys.unsetenv(unset)
    set <- old_env[!is.na(old_env)]
    if (length(set)) do.call(Sys.setenv, as.list(set))
  }
)
elapsed <- as.numeric(Sys.time() - t0, units = "secs")

# --- read back and evaluate ----------------------------------------------
julia_report <- if (file.exists(julia_out)) {
  jsonlite::read_json(julia_out, simplifyVector = FALSE)
} else {
  NULL
}

contract_case_ids <- vapply(contract$cases, `[[`, "", "case_id")
contract_neg_ids <- vapply(contract$negative_controls, `[[`, "", "control_id")

results_ok <- !is.null(julia_report) &&
  identical(julia_status, 0L) &&
  identical(julia_report$status, "PASS") &&
  isTRUE(julia_report$all_positive_pass) &&
  isTRUE(julia_report$negative_controls_behaved_as_expected) &&
  isTRUE(julia_report$all_checks) &&
  identical(julia_report$case_count, length(contract$cases)) &&
  all(vapply(julia_report$cases, function(x) isTRUE(x$pass), logical(1)))

all_case_ids_seen <- if (!is.null(julia_report)) names(julia_report$cases) else character(0)
missing_case_ids <- setdiff(contract_case_ids, all_case_ids_seen)
extra_case_ids <- setdiff(all_case_ids_seen, contract_case_ids)
results_ok <- results_ok && length(missing_case_ids) == 0 && length(extra_case_ids) == 0

raw_lines <- vapply(contract_case_ids, function(id) {
  pass <- !is.null(julia_report) && id %in% names(julia_report$cases) &&
    isTRUE(julia_report$cases[[id]]$pass)
  paste(id, if (pass) "PASS" else "FAIL", "positive", sep = "\t")
}, character(1))
neg_lines <- if (!is.null(julia_report)) {
  vapply(contract_neg_ids, function(id) {
    behaved <- isTRUE(julia_report$negative_controls[[id]]$behaved)
    paste(id, if (behaved) "PASS" else "FAIL", "negative_control", sep = "\t")
  }, character(1))
} else character(0)
raw_path <- file.path(output_dir, "results.tsv")
writeLines(c(raw_lines, neg_lines), raw_path)

diag_lines <- character(0)
if (!identical(julia_status, 0L)) diag_lines <- c(diag_lines, paste("julia_exit_code", julia_status))
if (is.null(julia_report)) diag_lines <- c(diag_lines, "julia-results.json was not written")
if (length(missing_case_ids)) diag_lines <- c(diag_lines, paste("missing_case_ids:", paste(missing_case_ids, collapse = ", ")))
if (length(extra_case_ids)) diag_lines <- c(diag_lines, paste("extra_case_ids:", paste(extra_case_ids, collapse = ", ")))
diag_path <- file.path(output_dir, "diagnostics.log")
writeLines(diag_lines, diag_path)

receipt <- list(
  status = if (isTRUE(results_ok)) "PASS" else "FAIL",
  scope = "CORE070_AGHQ_CONTROL_BATCH",
  reference_commit = contract$reference_commit,
  contract_sha256 = contract_sha256,
  source_pins = contract$source_pins,
  source_unchanged = TRUE,
  case_count = contract$expected_case_count,
  negative_control_count = length(contract$negative_controls),
  expected_case_ids = contract_case_ids,
  negative_control_case_ids = contract_neg_ids,
  julia_exit_code = julia_status,
  julia_elapsed_seconds = elapsed,
  julia_results_sha256 = if (file.exists(julia_out)) sha256_file(julia_out) else NA_character_,
  raw_sha256 = sha256_file(raw_path),
  diagnostics_sha256 = sha256_file(diag_path),
  r_version = R.version.string,
  gllvmTMB_version = as.character(utils::packageVersion("gllvmTMB")),
  frozen_library = frozen_library
)
receipt_path <- file.path(output_dir, "receipt.json")
jsonlite::write_json(receipt, receipt_path, auto_unbox = TRUE, pretty = TRUE, null = "null")

cat("CORE070_AGHQ_CONTROL_BATCH_", receipt$status, "\n", sep = "")
quit(status = if (isTRUE(results_ok)) 0L else 1L)
